feat(gitlab) Implement push & merge_request hooks for gitlab

markstory opened the pull request:
Tweak how we handle gitlab external_id and webhook secrets. GitLab push events do not provide much context for us to locate the sentry organization and repository. We need to leverage the only field we have (the webhook secret) to carry an identifier that lets us locate the integration installation. With this and the rest of the push payload we can locate the repository to sync commits in.

This does not handle pushes that contain more than 20 commits. Like GitHub we only sync the first 20 commits of any push. The remaining diff gets collected when a Release is created.

I've also made some changes around project id/project slug that aim to normalize on project id. Slugs can be changed in gitlab and require clunky URL encoding to work with.

Refs APP-597

lauryndbrown commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -79,92 +136,68 @@ def create_commits(self, event, organization, repo):
             return HttpResponse(status=400)
 
-        handler()(organization, event)
+        handler()(integration, integration.organizations.first(), event)
```
So, when you use `first` that is the equivalent of querying and then using an order by. I can't actually recall how organizations are ordered, but I have a feeling the default behavior isn't probably what you would want. 

https://django.readthedocs.io/en/1.6.x/ref/models/querysets.html?#first 

Is there a reason the organizations are being sorted, or do you simply just want the first element?

markstory replied:
I'm using `.first()` as `integration.organizations` is a Many To Many association and I couldn't use `[0]` on the property because `ManyToManyManager` doesn't support that interface. I'm not interested in the order, I just want one record from the relation.

lauryndbrown commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -79,92 +136,68 @@ def create_commits(self, event, organization, repo):
             return HttpResponse(status=400)
 
-        handler()(organization, event)
+        handler()(integration, integration.organizations.first(), event)
```
Why only pick the first organization associated with the integration to create the `PullRequest` or `Commit` objects? That is going to run into a couple of issues:
- Because GitLab uses default identity (Oauth only), all the auth information is supplied by the user. If, for some reason, the default auth in the first organization doesn't work (the user deletes their account or something of that nature), this will end the attempt to update. But if you had tried the second organization's auth  it would have succeeded, you fail to update the information altogether.
- Second, if there are other organizations associated with the integration, you are not updating them, as both `PullRequest` and `Commit` model objects are scoped via an `organization_id`. You will need to create those objects for all orgs associated with an integration for the behavior to work as expected.

markstory replied:
> Second, if there are other organizations associated with the integration

Is this possible? I observed locally that each organization that connected the gitlab integration made a new `sentry_integration` record in the database. Each of these records only ever had 1 record in the join table. I'm assuming the ManyToMany association is either for other integrations or old behavior.

lauryndbrown replied:
Yes it is possible and expected. I would suggest not looking at Github as the example, in many ways it is a special case. In GitHub Apps there is no way to have a many to many relationship. This is a restraint on their end, but not on ours. 

The many to many relationship is intended. The idea is that when you install the same account to a different organization all that has to be added is another orangizationIntegration model. 
And the integration itself can be shared across organizations that way.

So for all other integrations but GitHub, the many to many relationship is expected behavior and must be handled here. The closest integration to GitLab is AzureDevOps due to the default auth. You will run into very similar issues I did when creating that integration. But they are issues that must be handled.

markstory replied:
> The idea is that when you install the same account to a different organization all that has to be added is another orangizationIntegration model. And the integration itself can be shared across organizations that way.

I installed the current GitLab integration into different sentry organizations as the same user and as different ones, and each install created new `sentry_integration` records. I wasn't able to reproduce an integration being shared by multiple organizations. The OrganizationIntegrationDetails deletion logic doesn't look like it would correctly handle integrations being shared across organizations either.

https://github.com/getsentry/sentry/blob/d125ced4f403086d03a6994a4b43d534827b9c4c/src/sentry/api/endpoints/organization_integration_details.py#L30-L56

A delete in one organization would delete the integration from all the other organizations as well.

markstory replied:
I was able to find multiple organizationintegrations for slack, jira and bitbucket. I don't understand how that state is created, but I'll update the gitlab implementation so that if does happen it behaves correctly.

markstory commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -22,27 +28,79 @@
+                repository_id=repo.id,
+                key=number,
+                values={
+                    'organization_id': organization.id,
```
Why not? In the create mode it is a necessary foreign key.

markstory commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -79,92 +136,68 @@ def create_commits(self, event, organization, repo):
+            integration = Integration.objects.filter(
+                provider=self.provider,
+                external_id=request.META['HTTP_X_GITLAB_TOKEN']
+            ).prefetch_related('organizations').get()
```
Integration.organizations is a ManyToMany relation. `select_related` emits a `JOIN` which selects redundant rows when the assoication has multiple records.

markstory commented:
@lauryndbrown Thank you for pointing out the external_id issue, I completely forgot that we had talked about that. I think I've addressed all the feedback you gave.

lauryndbrown commented on src/sentry/integrations/gitlab/client.py:
```diff
@@ -156,28 +150,23 @@ def search_issues(self, query):
-    def create_project_webhook(self, project):
-        path = GitLabApiClientPath.project_hooks.format(
-            project=quote(project, safe=''))
+    def create_project_webhook(self, project_id):
```
@markstory sounds good thanks for looking into it

lauryndbrown commented:
@markstory I wanted to comment further on the many to many relationship with integrations. It's important to understand. For two organizations to share an integration, it  must have the same provider (gitlab) and the same `external_id`. The `external_id`  was a combination of the `base_url` and the `group_id`. If GitLab creates that same `external_id`, they will be the same integration. You can see where this happens here: https://github.com/getsentry/sentry/blob/master/src/sentry/integrations/pipeline.py#L32:L46

The reason you're not seeing this behavior, is because you changed the `external_id` for gitlab. https://github.com/getsentry/sentry/pull/10093/files#diff-0170465b9df132e1186667d57fc315a6R258 This is not what you want in this case. The group and base url is a better `external_id` given that when the same integration is create it will be reused. If you could please change that so that the models work as intended.

lauryndbrown commented on src/sentry/integrations/gitlab/integration.py:
```diff
@@ -248,23 +246,27 @@ def build_integration(self, state):
+            # We splice the host & secret together to create an external id.
+            # This id is used as a webhook secret so we can find the matching
+            # sentry org later on.
+            'external_id': u'{}:{}'.format(hostname, uuid4().hex),
```
Should be hostname + group_id instead

markstory replied:
That would also be used as the 'secret' for webhooks then. GitLab webhooks don't contain any other data that lets us track back to the sentry organization.

lauryndbrown commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -79,92 +142,79 @@ def create_commits(self, event, organization, repo):
+            logger.info(
+                'gitlab.webhook.invalid-json',
+                extra={
+                    'external_id': integration.external_id
```
I'm not sure if this is too much, but would you add the request body here?

markstory replied:
The payload could be pretty big and I would be concerned with clogging up the pipe to Elasticsearch. Having the external_id should give enough information to get close to the problematic hook in GitLab.

lauryndbrown commented on src/sentry/integrations/gitlab/webhooks.py:
```diff
@@ -79,92 +142,79 @@ def create_commits(self, event, organization, repo):
-            logger.error(
-                PROVIDER_NAME + '.webhook.missing-body', extra={
-                    'organization_id': organization.id,
+        if not constant_time_compare(secret, integration.metadata['webhook_secret']):
```
This is cool. I hadn't seen this before. Maybe we could add this to the others as well?

markstory replied:
Yes. I'll add an issue in jira.
